tiledb/sm/array/array_directory.cc

```diff
@@ -577,15 +569,59 @@ bool ArrayDirectory::loaded() const {
 /*         PRIVATE METHODS           */
 /* ********************************* */
 
-tuple<Status, optional<std::vector<URI>>> ArrayDirectory::list_root_dir_uris() {
-  // List the array directory URIs
-  auto timer_se = stats_->start_timer("list_root_uris");
+const std::set<const std::string>& ArrayDirectory::dir_names() {
+  static const std::set<const std::string> dir_names = {
+      constants::array_schema_dir_name,
+      constants::array_metadata_dir_name,
+      constants::array_fragment_meta_dir_name,
+      constants::array_fragments_dir_name,
+      constants::array_commits_dir_name,
+      constants::array_dimension_labels_dir_name};
+
+  return dir_names;
+}
+
+std::vector<URI> ArrayDirectory::ls(const URI& uri) const {
+  auto&& [st, opt_dir_entries] = resources_.get().vfs().ls_with_sizes(uri);
+  throw_if_not_ok(st);
+  auto dir_entries = opt_dir_entries.value();
+  auto dirs = dir_names();
+  std::vector<URI> uris;
 
-  std::vector<URI> array_dir_uris;
-  RETURN_NOT_OK_TUPLE(
-      resources_.get().vfs().ls(uri_, &array_dir_uris), nullopt);
+  for (auto entry : dir_entries) {
+    auto entry_uri = URI(entry.path().native());
+
+    // Filter out empty files of the same name as the directory
+    if (entry_uri.remove_trailing_slash() == uri.remove_trailing_slash() &&
+        entry.file_size() == 0) {
+      continue;
+    }
+
+    // Always list directories
+    if (entry.is_directory()) {
+      uris.emplace_back(entry_uri);
+      continue;
+    }
```

Comment:
Lets move this above the zero byte file check. I don't think we'll ever hit the case where we get the directory name from an `ls`, but moving this check means the file check is guaranteed to work only on files.